### ISSUE TITLE
Investigate application <-> api relationship in auth0 and make sure I am following secure practices

### ISSUE THREAD

**Issue by joshuaauerbachwatson** [joshuaauerbachwatson/unigame#9]
I've tried several times to fix the token expiration at one week rather than one day.   But, it seems like a login is required daily, which I think is inconvenient and not warranted by the pretty limited value of the assets being protected on the server.

**Comment by joshuaauerbachwatson** [comment]
After several forays into Auth0 documentation:

- I still don't _really_ understand how multiple applications can use the same API but it does seem like a common pattern
- Because the server uses Auth0 it will necessary for the client library to use Auth0-issued tokens.
- However, because the callback URIs are tied to the native bundle id, each distinct native app will need its own Auth0 application.
- Exactly what this means for the "openness" of the framework is unclear.  At least I should be able to use it for multiple apps with my own Auth0 account.   It is possible that an unrelated person wishing to build on `unigame` would need to run their own server.

Because `unigame-core` will shortly stop being an app (at least I hope that works), we can delay confronting this problem for a while.   In `unigame-core` we just need to abstract away the Auth0 details.   We can probably get away with no Auth0 code dependency with just a "get new token" method.  Just the token and its expiration date needs to be returned.

**Comment by joshuaauerbachwatson** [comment]
> * It is possible that an unrelated person wishing to build on `unigame` would need to run their own server.

As noted recently in #23,  I could (in the very short run) create new apps on request for third parties wishing to create `unigame` apps.  But, it would really be a problem if I couldn't set up the server to accept tokens issued by different `Auth0` accounts.   This still needs exploration and experimentation (and perhaps consultation with Auth0 people).

> We can probably get away with no Auth0 code dependency with just a "get new token" method.

That's what I actually did in developing a version of `AnyCards` that uses `unigame`.   But, it means that every app needs its own `Auth0` code dependency and code (duplicated in every app) to use it.

A better design would be to restore the `Auth0` dependency in `unigame` and move the logic to do the login back there.  Each app, then must just provide its own `Auth0.plist` with a cliientID and domain but otherwise can avoid Auth0-specific code.   The only hard part of this is testing it.  Testing it with `AnyCards` will probably be easy (unless I find some gotcha) but then it has to be tested with a fresh bundle id to be meaningful.   To test it with a new Auth0 domain means (I think) creating a second Auth0 account with some variation on my identity ... not sure I can do that.

**Comment by joshuaauerbachwatson** [comment]
I believe I now understand the issues pretty well.

There is no _direct_ tie between the `Auth0` application used to issue access tokens and the `Auth0` API which consumes them.  _However_:

1.  The `iss` field of the token gives the domain, which is basically the `Auth0` account.  When the server validates a token, it checks the domain.   So, without further tweaks to the server, tokens issued by a different account will fail.
2. On the other hand, tokens issued by the same account, regardless of application, will succeed.  Thus, as long as I am the one setting up new applications for new apps, all will be well.
3. The `token` also includes the `aud` (audience) field which is also checked at the server.   As far as I can see, the audience value is specified at login and not contrained by `Auth0` (that is, within a domain, any application can request any audience.

So, unless I figure out something more complex and subtle, a completely unrelated party cannot write an app using `unigame` and sharing the same server with my apps.   Auth0 doc seems to be saying that yes, you can have more than one domain accessing an API but, no, those domains cannot be defined in separate Auth0 tenants.   Whether they really mean "cannot" or just "it's not a good idea" I'm not sure but clearly you will be twisting the intended use if you try to violate this rule.  Also, clearly, the server must know about all the domains it is checking for so, at a minimum, an unrelated Auth tenant wanting to use my server would have to communicate this information.

On the other hand, since all of this is open source, such a party could easily spin up his own `Auth0` account and his own server and by totally in business.

In summary:

1.  Anyone wanting to share a server with me will have to send me the information needed for me to create an application on his behalf.   I would then send that person the domain and client id to use for his app.
2. When I get tired of (1) (which is likely to happen quickly if there is any uptake), people who want to use `unigame` would need to run their own server and set up their own `Auth0` tenancy.   They could use the same audience or not, since I think audiences are effectively scoped within domains.

I think that concludes the usefulness of this issue.
